src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/JavaSourceBuilder.java

```diff
@@ -205,7 +205,14 @@ String build() {
      */
     final String uniqueNestedClassName(String name) {
         name = Utils.javaSafeIdentifier(name);
-        return nestedClassNames.add(name.toLowerCase()) ? name : (name + "$" + nestedClassNameCount++);
+        var notSeen = nestedClassNames.add(name.toLowerCase());
+        var notEnclosed = !isEnclosedBySameName(name.toLowerCase());
+        return notSeen && notEnclosed? name : (name + "$" + nestedClassNameCount++);
+    }
+
+    // is the name enclosed enclosed by a class of the same name?
+    boolean isEnclosedBySameName(String name) {
+        return false;
```

Comment:
Should we add className().equals(name) in here - just to be on the safe side>

I guess it's never possible for a nested class to have the same name of a "toplevel" class (e.g. generated by HeaderFileBuilder), because we explicitly mangle the names (e.g. `toplevel$nested`). But I think adding the name check at the toplevel is the correct thing to do code-wise.

Then in the subclass you can do `super.isEnclosedBySameName(name) || enclosing.isEnclosedBySameName(name))`